tests/unit/s2n_record_test.c

```diff
@@ -429,6 +429,37 @@ int main(int argc, char **argv)
         EXPECT_EQUAL(s2n_stuffer_data_available(&conn->header_in), 3);
     };
 
+    /* Record version is recorded for the first message received (Client Hello) */
+    {
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        uint8_t content_type = 0;
+        uint16_t fragment_length = 0;
+        uint8_t out = 0;
+        uint8_t header[5] = { 0x16, /* Record type */
+            0x03, 0x01,             /* Protocol version: TLS10 */
+            0x00, 0x00 };           /* Record size */
+
+        uint8_t altered_header[5] = { 0x16, /* Record type */
+            0x03, 0x02,                     /* Protocol version: TLS11 */
+            0x00, 0x00 };                   /* Record size */
+
+        EXPECT_SUCCESS(s2n_stuffer_write_bytes(&server_conn->header_in, header, sizeof(header)));
+        EXPECT_SUCCESS(s2n_record_header_parse(server_conn, &content_type, &fragment_length));
+        /* Record TLS version is retrieved as written in the header */
+        EXPECT_EQUAL(server_conn->client_hello.legacy_record_version, S2N_TLS10);
+
+        server_conn->actual_protocol_version_established = 1;
+        EXPECT_SUCCESS(s2n_stuffer_wipe(&server_conn->header_in));
+
+        EXPECT_SUCCESS(s2n_stuffer_write_bytes(&server_conn->header_in, altered_header, sizeof(header)));
+        /* This will fail as the protocol version is technically an error, however, we still want to check
+            * that the protocol version recorded was unchanged. */
+        EXPECT_FAILURE_WITH_ERRNO(s2n_record_header_parse(server_conn, &content_type, &fragment_length), S2N_ERR_BAD_MESSAGE);
```

Comment:
Kind of a nit - but is there a way to write this so that `s2n_record_header_parse` succeeds? Technically it could be failing before `legacy_record_version` is set, which could be the reason it's unmodified rather than the actual logic.